File: February 2025 GFG SOLUTION/February-05.c
//{ Driver Code Starts
// Initial Template for C
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* A binary tree node has data, pointer to left child
   and a pointer to right child */
typedef struct Node {
    int data;
    struct Node *left;
    struct Node *right;
} Node;

// Function to create a new node
Node *newNode(int x) {
    Node *node = (Node *)malloc(sizeof(Node));
    node->data = x;
    node->left = NULL;
    node->right = NULL;
    return node;
}

// Function to create a queue
typedef struct QueueNode {
    Node *treeNode;
    struct QueueNode *next;
} QueueNode;

typedef struct Queue {
    QueueNode *front, *rear;
} Queue;

Queue *createQueue() {
    Queue *q = (Queue *)malloc(sizeof(Queue));
    q->front = q->rear = NULL;
    return q;
}

void enqueue(Queue *q, Node *node) {
    QueueNode *newNode = (QueueNode *)malloc(sizeof(QueueNode));
    newNode->treeNode = node;
    newNode->next = NULL;
    if (q->rear == NULL) {
        q->front = q->rear = newNode;
        return;
    }
    q->rear->next = newNode;
    q->rear = newNode;
}

Node *dequeue(Queue *q) {
    if (q->front == NULL)
        return NULL;
    QueueNode *temp = q->front;
    Node *node = temp->treeNode;
    q->front = q->front->next;
    if (q->front == NULL)
        q->rear = NULL;
    free(temp);
    return node;
}

int isQueueEmpty(Queue *q) {
    return q->front == NULL;
}

// Function to Build Tree
Node *buildTree(char *str) {
    // Corner Case
    if (strlen(str) == 0 || str[0] == 'N')
        return NULL;

    // Tokenize the input string
    char *token = strtok(str, " ");
    Node *root = newNode(atoi(token));

    Queue *q = createQueue();
    enqueue(q, root);

    while (!isQueueEmpty(q)) {
        Node *currNode = dequeue(q);

        // Left child
        token = strtok(NULL, " ");
        if (token == NULL)
            break;
        if (strcmp(token, "N") != 0) {
            currNode->left = newNode(atoi(token));
            enqueue(q, currNode->left);
        }

        // Right child
        token = strtok(NULL, " ");
        if (token == NULL)
            break;
        if (strcmp(token, "N") != 0) {
            currNode->right = newNode(atoi(token));
            enqueue(q, currNode->right);
        }
    }

    free(q);
    return root;
}

// Generate output as level order of tree
char *levelOrder(Node *root) {
    if (root == NULL)
        return strdup("N\n");

    char *str = (char *)malloc(2000000 * sizeof(char));
    str[0] = '\0';
    Queue *q = createQueue();
    enqueue(q, root);

    while (!isQueueEmpty(q)) {
        Node *curr = dequeue(q);

        if (curr == NULL) {
            strcat(str, "N ");
            continue;
        }

        char buffer[12];
        sprintf(buffer, "%d ", curr->data);
        strcat(str, buffer);

        enqueue(q, curr->left);
        enqueue(q, curr->right);
    }

    // Trim trailing non-numeric characters
    int len = strlen(str);
    while (len > 0 && !isdigit(str[len - 1])) {
        str[--len] = '\0';
    }

    free(q);
    return str;
}


// } Driver Code Ends
// User function Template for C
/*
typedef struct Node {
    int data;
    struct Node *left;
    struct Node *right;
} Node;

Node *newNode(int x) {
    Node *node = (Node *)malloc(sizeof(Node));
    node->data = x;
    node->left = NULL;
    node->right = NULL;
    return node;
}
*/
void mirror(Node *n) {
    if (!n) return;
    mirror(n->left);
    mirror(n->right);
    Node* t = n->left;
    n->left = n->right;
    n->right = t;
}


//{ Driver Code Starts.

int main() {
    int tc;
    scanf("%d ", &tc);

    while (tc--) {
        char str[2000000];
        fgets(str, sizeof(str), stdin);
        str[strlen(str) - 1] = '\0';

        Node *root = buildTree(str);
        mirror(root);

        char *result = levelOrder(root);
        puts(result);
        // printf("\n");
        printf("~\n");

        free(result);
    }

    return 0;
}

// } Driver Code Ends

File: February 2025 GFG SOLUTION/February-05.py
#User function Template for python3

'''
class Node:
    def _init_(self, val):
        self.right = None
        self.data = val
        self.left = None
'''
# your task is to complete this function

# class Solution:
#     def mirror(self, root):
#         if not root: return
#         self.mirror(root.left)
#         self.mirror(root.right)
#         root.left, root.right = root.right, root.left

class Solution:
    def mirror(self, root):
        if not root:
            return
        self.mirror(root.left)
        self.mirror(root.right)
        root.left, root.right = root.right, root.left
#{ 
 # Driver Code Starts
#Initial Template for Python 3

#Contributed by Sudarshan Sharma
from collections import deque


# Tree Node
class Node:

    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


from collections import deque


class Node:

    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


def levelOrder(root):
    if root is None:
        return "N\n"

    result = []
    qq = deque()
    qq.append(root)

    while qq:
        curr = qq.popleft()

        if curr is None:
            result.append("N")
            continue

        result.append(str(curr.data))
        qq.append(curr.left)
        qq.append(curr.right)

    # Remove trailing non-numeric elements
    while result and not result[-1].isnumeric():
        result.pop()

    return " ".join(result)


def inorderTraversal(root):
    # Code here
    inorderTraversalUtil(root)
    print()


# Function to Build Tree
def buildTree(s):
    #Corner Case
    if (len(s) == 0 or s[0] == "N"):
        return None

    # Creating list of strings from input
    # string after spliting by space
    ip = list(map(str, s.split()))

    # Create the root of the tree
    root = Node(int(ip[0]))
    size = 0
    q = deque()

    # Push the root to the queue
    q.append(root)
    size = size + 1

    # Starting from the second element
    i = 1
    while (size > 0 and i < len(ip)):
        # Get and remove the front of the queue
        currNode = q[0]
        q.popleft()
        size = size - 1

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if (currVal != "N"):

            # Create the left child for the current node
            currNode.left = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.left)
            size = size + 1
        # For the right child
        i = i + 1
        if (i >= len(ip)):
            break
        currVal = ip[i]

        # If the right child is not null
        if (currVal != "N"):

            # Create the right child for the current node
            currNode.right = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.right)
            size = size + 1
        i = i + 1
    return root


if __name__ == "__main__":
    t = int(input())
    for _ in range(0, t):
        s = input()
        root = buildTree(s)
        Solution().mirror(root)
        print(levelOrder(root))
        print("~")

# } Driver Code Ends
